core/src/cubos/debug_render.cpp

```diff
@@ -0,0 +1,183 @@
+
+    gl::VertexArrayDesc vaDesc;
+    vaDesc.elementCount = 1;
+    vaDesc.elements[0].name = "position";
+    vaDesc.elements[0].type = gl::Type::Float;
+    vaDesc.elements[0].size = 3;
+    vaDesc.elements[0].buffer.index = 0;
+    vaDesc.elements[0].buffer.offset = 0;
+    vaDesc.elements[0].buffer.stride = 3 * sizeof(float);
+    vaDesc.buffers[0] = vb;
+    vaDesc.shaderPipeline = pipeline;
+    objCube.va = debugRenderDevice->createVertexArray(vaDesc);
+}
+
+void debug::init(io::Window* window)
+{
+    debugWindow = window;
+    debugRenderDevice = &debugWindow->getRenderDevice();
+
+    auto vs = debugRenderDevice->createShaderStage(gl::Stage::Vertex, R"(
+            #version 330 core
+
+            in vec3 position;
+
+            uniform MVP
+            {
+                mat4 mvp;
+            };
+
+            void main()
+            {
+                gl_Position = mvp * vec4(position, 1.0f);
+            }
+        )");
+
+    auto ps = debugRenderDevice->createShaderStage(gl::Stage::Pixel, R"(
+            #version 330 core
+
+            out vec4 color;
+
+            uniform vec3 objColor;
+
+            void main()
+            {
+                color = vec4(objColor, 1.0f);
+            }
+        )");
+
+    pipeline = debugRenderDevice->createShaderPipeline(vs, ps);
+
+    initCube();
+
+    mvpBuffer = debugRenderDevice->createConstantBuffer(sizeof(glm::mat4), nullptr, gl::Usage::Dynamic);
+    mvpBindingPoint = pipeline->getBindingPoint("MVP");
+    colorBindingPoint = pipeline->getBindingPoint("objColor");
+
+    RasterStateDesc rsDesc;
+
+    rsDesc.rasterMode = RasterMode::Fill;
+    FillRasterState = debugRenderDevice->createRasterState(rsDesc);
+
+    rsDesc.rasterMode = RasterMode::Wireframe;
+    WireframeRasterState = debugRenderDevice->createRasterState(rsDesc);
+}
+
+void debug::drawCube(glm::vec3 center, glm::vec3 size, float time, glm::vec3 color)
+{
+    requests.push_back(
+        DebugDrawRequest{objCube, FillRasterState, glm::translate(center) * glm::scale(size), time, color});
+}
+
+void debug::drawWireCube(glm::vec3 center, glm::vec3 size, float time, glm::vec3 color)
+{
+    requests.push_back(
+        DebugDrawRequest{objCube, WireframeRasterState, glm::translate(center) * glm::scale(size), time, color});
+}
+
+void debug::drawSphere(glm::vec3 center, float radius, float time, glm::vec3 color)
+{
+    requests.push_back(DebugDrawRequest{objSphere, FillRasterState,
+                                        glm::translate(center) * glm::scale(glm::vec3(radius)), time, color});
+}
+
+void debug::drawWireSphere(glm::vec3 center, float radius, float time, glm::vec3 color)
+{
+    requests.push_back(DebugDrawRequest{objSphere, WireframeRasterState,
+                                        glm::translate(center) * glm::scale(glm::vec3(radius)), time, color});
+}
+
+void debug::flush(glm::mat4 vp, double deltaT)
+{
+    debugRenderDevice->setShaderPipeline(pipeline);
```

Comment:
This should be made thread safe: add a static std::mutex to the class

Reply:
which methods should be thread safe exactly? the draw*() methods? I see no reason for the flush function to be called more than once per frame

Reply:
The draw and flush methods should be thread safe: you could have another thread running adding objects to the debug renderer at the same time as the previous frame is being rendered or somethint. If we don't make it thread safe, it should be made explicit in the docs.